### PR TITLE
Store either "someexp" or "-someexp" in cookie, never both. 

### PR DESCRIPTION
**Pull request by lannka**
This is to fix the duplicated experiment entries in cookie as mentioned [here](https://github.com/ampproject/amphtml/pull/6709#discussion_r93081641) .

It's done by deserializing the experiment cookie into an object instead of an array:
`AMP_EXP=e1,-e2,e3` => 
```
{
  e1: true,
  e2: false,
  e3: true,
}
```

### CODE REVIEW

File: src/experiments.js [hunk 1]
```diff
@@ -185,25 +172,44 @@ export function toggleExperiment(win, experimentId, opt_on,
 /**
  * Returns a set of experiment IDs currently on.
  * @param {!Window} win
- * @return {!Array<string>}
+ * @return {!Object<string, boolean>}
+ * @visibleForTesting
  */
-function getExperimentIds(win) {
+export function getExperimentTogglesFromCookie(win) {
```
Comment (dvoytenko) [inline src/experiments.js]: Lately, I've been directing to keep the original method private to the module, but instead creating a specifically test method, e.g.
```
function getExperimentTogglesFromCookie(win) {
  ...
 ..
  ..
}

export function getExperimentToglesFromCookieForTesting(win) {
  return getExperimentTogglesFromCookie(win);
}
```

"xForTesting" methods are stripped, so we get a better protection.

File: src/experiments.js [hunk 2]
```diff
@@ -161,21 +153,16 @@ function calcExperimentOn(win, experimentId) {
...
-  const on = opt_on !== undefined ? opt_on : !currentlyOn;
+  const currentlyOn = isExperimentOn(win, experimentId);
+  const on = !!(opt_on !== undefined ? opt_on : !currentlyOn);
   if (on != currentlyOn) {
-    if (on) {
-      experimentIds.push(experimentId);
-      toggles[experimentId] = true;
-    } else {
-      experimentIds.splice(experimentIds.indexOf(experimentId), 1);
-      toggles[experimentId] = false;
```
Comment (erwinmombay) [inline src/experiments.js]: @zhouyx this should fix the bug

File: src/experiments.js [hunk 3]
```diff
@@ -161,21 +153,16 @@ function calcExperimentOn(win, experimentId) {
...
-    }
+    const toggles = experimentToggles();
+    toggles[experimentId] = on;
+
+    const cookieToggles = getExperimentTogglesFromCookie(win);
+    cookieToggles[experimentId] = on;
```
Comment (jridgewell) [inline src/experiments.js]: This means `opt_transientExperiment` stores could be persisted if anything after it persists.

File: src/experiments.js [hunk 4]
```diff
@@ -124,18 +124,10 @@ export function isExperimentOnAllowUrlOverride(win, experimentId) {
...
-  if (disableFlag) {
-    return false;
-  }
-
-  // Enabling cookie flag.
-  const cookieFlag = experiments.indexOf(experimentId) != -1;
-  if (cookieFlag) {
-    return true;
+  if (cookieToggles[experimentId] !== undefined) {
+    return cookieToggles[experimentId];
```
Comment (jridgewell) [inline src/experiments.js]: This won't reflect transient experiment toggles, now.

Comment (lannka) [inline src/experiments.js]: as talked offline, it works fine. I added a test to prove.

File: src/experiments.js (hunk 1 above)
Comment (lannka) [inline src/experiments.js]: sounds cool. done.

**Pull request merged**